distributed/event.py

```diff
@@ -0,0 +1,251 @@
+    we keep a mapping from name to an asyncio.Event and call
+    every function (wait, set, clear, is_set) directly on these
+    events.
+
+    However, this would cause a memory leak: created events in the
+    dictionary are never removed.
+    For this, we also keep a counter for the number of waiters on
+    a specific event.
+    If an event is set, we need to keep track of this state so
+    we can not remove it (the default flag is false).
+    If it is unset but there are waiters, we can also not remove
+    it, as those waiters would then have dangling futures.
+    Therefore the only time we can remove the event from our dict
+    is when the number of waiters is 0 and the event flag is cleared.
+    """
+
+    def __init__(self, scheduler):
+        self.scheduler = scheduler
+        # Keep track of all current events, identified by their name
+        self._events = dict()
+        # Keep track on how many waiters are present, so we know when
+        # we can remove the event
+        self._waiter_count = defaultdict(int)
+
+        self.scheduler.handlers.update(
+            {
+                "event_wait": self.event_wait,
+                "event_set": self.event_set,
+                "event_clear": self.event_clear,
+                "event_is_set": self.event_is_set,
+            }
+        )
+
+        self.scheduler.extensions["events"] = self
+
+    async def event_wait(self, stream=None, name=None, timeout=None):
+        """ Wait until the event is set to true.
+        Returns false, when this did not happen in the given time
+        and true otherwise.
+        """
+        with log_errors():
+            name = self._normalize_name(name)
+
+            event = self._get_or_create_event(name)
+            future = event.wait()
+            if timeout is not None:
+                future = asyncio.wait_for(future, timeout)
+
+            self._waiter_count[name] += 1
+            try:
+                await future
+            except TimeoutError:
+                return False
+            finally:
+                self._waiter_count[name] -= 1
+
+                if not self._waiter_count[name] and not event.is_set():
+                    # No one is waiting for this
+                    # and as the default flag for an event is false
+                    # we can safely remove it
+                    self._delete_event(name)
+
+            return True
+
+    async def event_set(self, stream=None, name=None):
+        """ Set the event with the given name to true.
+
+        All waiters on this event will be notified.
+        """
+        with log_errors():
+            name = self._normalize_name(name)
+            # No matter if someone is listening or not,
+            # we set the event to true
+            event = self._get_or_create_event(name)
+            self.scheduler.loop.add_callback(event.set)
+
+    async def event_clear(self, stream=None, name=None):
```

Comment:
It looks like this function doesn't have any `await`s in it, so we can probably make it a normal function.
```suggestion
    def event_clear(self, stream=None, name=None):
```